### ISSUE TITLE
 Mac support with M1 processor (darwin, arm64)

### ISSUE THREAD

**Issue by victorpaleologue** [KCL-Planning/VAL#56]
When using the VSCode PDDL extension, I am given the opportunity to download and configure a pre-built VAL tool. The matching binary is not found for my platform: Mac OS on an M1 processor (darwin, arm64).
I reported the issue there https://github.com/jan-dolejsi/vscode-pddl/issues/127, however I feel like this issue is that the VAL release for that platform is missing.

**Comment by victorpaleologue** [comment]
Can I help? I am at ease with CMake and GitHub Actions.
I would have tried out updating the Azure pipeline, but I cannot find the relevant YAML file.
If someone can point at where to start, I could do something.

**Comment by DerekLong101** [comment]
Ah, that is possible. We checked compilation on a Mac, but I don't think any of us uses one as a matter of course, so we don't tend to be good at keeping up with them. So, yes, we could probably benefit from some help. The Azure pipeline is inside the company firewalls, so I don't think it will be easy to give access. But we might be able to give you the scripts we use so that you can do the surgery and send them back for us. Let me check.
________________________________
From: Victor Paléologue ***@***.***>
Sent: 16 June 2022 16:43
To: KCL-Planning/VAL ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [KCL-Planning/VAL] Mac support with M1 processor (darwin, arm64) (Issue #56)


When using the VSCode PDDL extension, I am given the opportunity to download and configure a pre-built VAL tool. The matching binary is not found for my platform: Mac OS on an M1 processor (darwin, arm64).
I reported the issue there jan-dolejsi/vscode-pddl#127<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fjan-dolejsi%2Fvscode-pddl%2Fissues%2F127&data=05%7C01%7Cderek.long%40kcl.ac.uk%7C7ea983be626f4c13649b08da4faef7c7%7C8370cf1416f34c16b83c724071654356%7C0%7C0%7C637909910149592278%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=jgnEUBHZ3ceeXDlPm%2F7AtJRIhGItMbTs69MoPyXc9bI%3D&reserved=0>, however I feel like this issue is that the VAL release for that platform is missing.

**Comment by victorpaleologue** [comment]
FYI it compiles well on Mac, all we need is a release.
I'm still available for help.

**Comment by DerekLong101** [comment]
Thanks, Victor. I will definitely come back to you on this - I am currently submerged with a few other jobs. This is on the stack!
________________________________

**Comment by jan-dolejsi** [comment]
> Can I help? I am at ease with CMake and GitHub Actions. I would have tried out updating the Azure pipeline, but I cannot find the relevant YAML file. If someone can point at where to start, I could do something.

Yes, you can definitely help. Modifying the azure pipeline to harvest and publish the binary artifacts would be the simplest change at the end. The hard part is to change the build scripts, so they produce the arm binaries. You can see the `azure-pipeline.yml` in the root of the repo. It is simply calling `./scripts/build_linux_windows.sh` and `./scripts/build_macos.sh` on ubuntu and MacOS respectively. The arm64 script should be added either to the `build_linux_windows.sh`, if the compilation for arm64 can be done on linux, or as a separate `build_arm64.sh` script, which we would add to the pipeline to run on hosted arm64 pool (not sure it exists).

The pipeline then runs over there: https://dev.azure.com/schlumberger/ai-planning-validation/_build?definitionId=2

**Comment by DerekLong101** [comment]
Thanks, Jan!
________________________________

**Comment by victorpaleologue** [comment]
> You can see the `azure-pipeline.yml` in the root of the repo.

Sorry but I cannot find that file. Could you provide me a link to it?

**Comment by jan-dolejsi** [comment]
Oh, my bad. I did not realize the [pipeline file was only in my fork](https://github.com/jan-dolejsi/VAL/blob/master/azure-pipelines.yml). Looking closer, it was work I started and did not finish. It only includes the build job for linux and windows (targeting linux build machine). But it would eventually include the MacOS job too. The [build pipeline](https://dev.azure.com/schlumberger/ai-planning-validation/_build?definitionId=2) that we currently run on Azure DevOps is using the infrastructure that predates yaml.

Nevertheless, what is needed for the M1 processor binaries is a build script first. Adding it to the pipeline would just mean calling the new script from using an additional task in the pipeline. Like what you can see on the [line 17](https://github.com/jan-dolejsi/VAL/blob/master/azure-pipelines.yml#L17).

**Comment by victorpaleologue** [comment]
Alright, I've put myself on it, but... [Microsoft-hosted Azure agents for Mac OS](https://docs.microsoft.com/en-us/azure/devops/pipelines/agents/hosted?view=azure-devops&tabs=yaml#hardware) do not seem to be running on M1 processors:
> Agents that run macOS images are provisioned on Mac pros with a 3 core CPU, 14 GB of RAM, and 14 GB of SSD disk space.

Knowing that M1 processors have many more CPU cores, this hardware is certainly x86_64. [GitHub runners have the same hardware specs](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources), btw, and they are explicit on the architecture being x86_64.

I checked `scripts/build_macos.sh` and it works fine already. But I won't be able to get your CI pipeline run it. Sorry, I genuinely hoped to help =/

However I can think of Conan and ConanCenter, which are dedicated to distribute portable native binaries to every platform. They do produce MacOS arm64 releases, and VAL could be distributed there too. Another big advantage of Conan is that if the binary is missing, it has access to all the recipes to rebuild it for the current host.
But that does not really fit the workflow of the PDDL plug-in for VSCode.

**Comment by jan-dolejsi** [comment]
@victorpaleologue are you saying that the script scripts/build_macos.sh produces binaries that you can run on the new shiny Mac?
Or only if you run the build locally, while the macos binaries that  come out of the pipeline are  not compatible? 

Yes the pipeline can be fixed eventually. Looks like Microsoft  is working on the sup6 for M1 build machines.

I am not  against an alternative build and download location. Ad long as the platform specific binaries can be downloaded from a fixed url. Ideally the version should ge part of the url too.

**Comment by victorpaleologue** [comment]
> @victorpaleologue are you saying that the script scripts/build_macos.sh produces binaries that you can run on the new shiny Mac? Or only if you run the build locally, while the macos binaries that come out of the pipeline are not compatible?

Yes, running the script on a new shiny Mac builds a binary compatible with new shiny Macs. But the ones that come out of the pipeline, built on older-but-perhaps-still-shiny Macs, are not compatible.

> Yes the pipeline can be fixed eventually. Looks like Microsoft is working on the sup6 for M1 build machines.
> 
> I am not against an alternative build and download location. Ad long as the platform specific binaries can be downloaded from a fixed url. Ideally the version should ge part of the url too.

Ok. I could publish my release someplace that won't move, but building manually next versions would be annoying. IMO it has to be automated, and rely on a machine dedicated for that. So I'd rather wait for Microsoft's support...

**Comment by DerekLong101** [comment]
I think the shine has come off the Macs.
________________________________

**Comment by DavidKaub** [comment]
Any updates on this so far?

**Comment by victorpaleologue** [comment]
Azure support for arm64-powered MacOS has not changed.
I did not take time to produce my own release. I cannot tell when I can do this.
In the meantime you will have to rebuild it yourself, it should build fine.

**Comment by victorpaleologue** [comment]
Would you be interested by having it available through Conan?

**Comment by victorpaleologue** [comment]
Still no change. If someone wants a manual release, let me know.
I'll let this sit in the meantime.

**Comment by DavidKaub** [comment]
A manual release would be welcome! Thanks

**Comment by victorpaleologue** [comment]
There it is.
You need to unzip it somewhere and configure the PDDL extension to use the binary called `Validate`.
[Val--Darwin.tar.gz](https://github.com/KCL-Planning/VAL/files/11180050/Val--Darwin.tar.gz)
[Val--Darwin.zip](https://github.com/KCL-Planning/VAL/files/11180051/Val--Darwin.zip)

**Comment by Asriel31** [comment]
my Norton antivirus have intercepted pdf file from pdf file from val--Darwin.tar.gz, it seems this file is not safe  :/

**Comment by Larox** [comment]
Hello there!

I'm trying to get the VAL tools working on 2026 and it seems the pipeline was not updated as planned... I just wanted to double-check if it is still in the works or we just should go with the binaries @victorpaleologue shared?

Writing this from M4 pro mac, its been a while since M1 came out 😆
